File: 1330.py
class Solution:
    def longestSubsequence(self, arr: List[int], difference: int) -> int:
        self.res = 0
        N = len(arr)
        @cache
        def helper(i, prev):
            if i >= N:
                return 0
            
            nc = helper(i+1, prev)
            count = 0
            if prev + difference == arr[i] or prev == -float('inf'):
                count = max(1 + helper(i+1, arr[i]), helper(i+1, prev))
            
            
            return max(nc,count)
        
        return helper(0, -float('inf'))

File: 1330.cpp
class Solution {
public:
    
    int longestSubsequence(vector<int>& arr, int difference) {
        map<int, int> c;
        int res = 0;
        for(auto i : arr){
            int before = c[i-difference] ? c[i-difference] : 0;
            c[i] = ++before;
            res = max(res, c[i]);
        }
        // for (auto [first, second] : c){
        //     cout<< first<<second<<endl;
        // }
        return res;
    }
};



